include/alpaka/math/Complex.hpp

```diff
@@ -0,0 +1,540 @@
+/* Copyright 2022 Sergei Bastrakov
+ *
+ * This file is part of alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <alpaka/core/Common.hpp>
+#include <alpaka/math/FloatEqualExact.hpp>
+
+#include <cmath>
+#include <complex>
+#include <iostream>
+
+namespace alpaka
+{
+    //! Implementation of a complex number useable on host and device.
+    //!
+    //! It follows the layout of std::complex and so array-oriented access.
+    //! Inside the class template implements all methods and operators as std::complex<T>.
+    //! Additionally, provides an implicit conversion to and from std::complex<T>.
+    //! All methods besides operators << and >> are host-device.
+    //! Note that it does not provide non-member functions of std::complex besides the operators.
+    //! Those are provided the same way as alpaka math functions for real numbers.
+    //!
+    //! Note that unlike most of alpaka, this is a concrete type template, and not merely a concept.
+    //!
+    //! Naming and order of the methods match https://en.cppreference.com/w/cpp/numeric/complex.
+    //!
+    //! @tparam T type of the real and imaginary part: float, double, or long double.
+    template<typename T>
+    class Complex
+    {
```

Comment:
```suggestion
    //! @tparam T type of the real and imaginary part: float, double, or long double.
    template<typename T>
    class Complex
    {
        static_assert(std::is_floating_point_v<T>);
```
Maybe enforce that `T` is a floating point in addition to the comment.